File: problems/268_missing_number/solution.cpp
// Leetcode 268. Missing Number
//
// Solution: Use XOR property.
//   - XOR of all numbers from 0 to n gives a value.
//   - XOR with all elements of the array cancels out the numbers that appear,
//     leaving the missing number.
//   This works because a ^ a = 0 and XOR is commutative and associative.
#include <vector>

class Solution {
 public:
  int missingNumber(const std::vector<int>& nums) {
    int n = static_cast<int>(nums.size());
    int xor_sum = 0;

    for (int i = 0; i <= n; ++i) {
      xor_sum ^= i;
    }

    for (const int num : nums) {
      xor_sum ^= num;
    }

    return xor_sum;
  }
};

File: problems/268_missing_number/Solution.hs
-- Leetcode 268. Missing Number

module Solution where

import Data.Bits (xor)
import Data.List (foldl')

missingNumber :: [Int] -> Int
missingNumber nums = foldl' xor xor_sum nums
  where n = length nums
        xor_sum = foldl' xor 0 [0..n]

missingNumber2 :: [Int] -> Int
missingNumber2 nums = foldl' xor 0 ([0..n] ++ nums)
  where n = length nums
